File: csharp/collatz-conjecture/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        if (number < 1)
        {
            throw new ArgumentOutOfRangeException("Only positive numbers are allowed");
        }

        return Steps(number, 0);
    }

    private static int Steps(int number, int steps)
    {
        if (number == 1)
        {
            return steps;
        }

        if (number % 2 == 0)
        {
            return Steps(number / 2, steps + 1);
        }

        return Steps(3 * number + 1, steps + 1);
    }
}

File: python/collatz-conjecture/collatz_conjecture.py
def steps(n: int) -> int:
    """
    Calculate the number of steps it takes to reach 1
    from the given value using the Collatz Conjecture.

    Args:
        n (int) : value to calculate the number of steps for

    Raises:
        ValueError: raised when the given value is out of bounds

    Returns:
        int: number of steps
    """

    if n <= 0:
        raise ValueError("Only positive integers are allowed")

    if n == 1:
        return 0

    if n % 2 == 0:
        return 1 + steps(n / 2)

    return 1 + steps(3 * n + 1)
